File: create-zoom-mail.mjs
#!/usr/bin/env node
/**
 * Zoom URL共有メールを Gmail API で下書き作成（既定）／送信（--send）。
 *
 * 認証ファイル:
 *   credentials/credentials.json … OAuth Desktop クライアント
 *   credentials/token.json … 初回認可後に自動保存
 *
 * 使い方:
 *   node create-zoom-mail.mjs              # 下書き作成
 *   node create-zoom-mail.mjs --dry-run    # 内容確認のみ
 *   node create-zoom-mail.mjs --test       # テスト宛先用の下書き
 *   node create-zoom-mail.mjs --send       # 送信（確認後のみ）
 */

import { createServer } from 'node:http';
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { createInterface } from 'node:readline';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { google } from 'googleapis';
import {
  BODY,
  CC,
  FROM,
  SUBJECT,
  TO,
} from './zoom_mail_payload.mjs';

const ROOT = dirname(fileURLToPath(import.meta.url));
const CRED_DIR = join(ROOT, 'credentials');
const CREDENTIALS_FILE = join(CRED_DIR, 'credentials.json');
const TOKEN_FILE = join(CRED_DIR, 'token.json');

const SCOPES_DRAFT = ['https://www.googleapis.com/auth/gmail.compose'];
const SCOPES_SEND = [
  'https://www.googleapis.com/auth/gmail.compose',
  'https://www.googleapis.com/auth/gmail.send',
];

function parseArgs(argv) {
  const args = {};
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a.startsWith('--')) {
      const key = a.slice(2);
      const next = argv[i + 1];
      args[key] = next && !next.startsWith('--') ? argv[++i] : 'true';
    }
  }
  return args;
}

function buildRawMessage({ to, subject, body, cc, from = FROM }) {
  const lines = [
    `From: ${from}`,
    `To: ${to}`,
    ...(cc ? [`Cc: ${cc}`] : []),
    `Subject: =?UTF-8?B?${Buffer.from(subject).toString('base64')}?=`,
    'MIME-Version: 1.0',
    'Content-Type: text/plain; charset=UTF-8',
    'Content-Transfer-Encoding: base64',
    '',
    Buffer.from(body, 'utf8').toString('base64'),
  ];
  return Buffer.from(lines.join('\r\n'))
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

async function authorize(scopes) {
  if (!existsSync(CREDENTIALS_FILE)) {
    console.error(
      `OAuth クライアントが見つかりません: ${CREDENTIALS_FILE}\n` +
        'Google Cloud Console で Desktop アプリの OAuth クライアント JSON を\n' +
        'credentials/credentials.json として配置してください。'
    );
    process.exit(2);
  }

  const raw = JSON.parse(readFileSync(CREDENTIALS_FILE, 'utf8'));
  const cfg = raw.installed || raw.web;
  if (!cfg) {
    console.error('credentials.json に installed / web クライアント情報がありません');
    process.exit(2);
  }

  const oAuth2Client = new google.auth.OAuth2(
    cfg.client_id,
    cfg.client_secret,
    'http://127.0.0.1:53682'
  );

  if (existsSync(TOKEN_FILE)) {
    oAuth2Client.setCredentials(JSON.parse(readFileSync(TOKEN_FILE, 'utf8')));
    try {
      if (oAuth2Client.isTokenExpiring?.() || !oAuth2Client.credentials.access_token) {
        const { credentials } = await oAuth2Client.refreshAccessToken();
        oAuth2Client.setCredentials(credentials);
        writeFileSync(TOKEN_FILE, JSON.stringify(credentials, null, 2));
      }
      return oAuth2Client;
    } catch {
      // fall through to interactive auth
    }
  }

  const authUrl = oAuth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: scopes,
    prompt: 'consent',
  });

  const code = await getAuthCode(authUrl);
  const { tokens } = await oAuth2Client.getToken(code);
  oAuth2Client.setCredentials(tokens);
  mkdirSync(CRED_DIR, { recursive: true });
  writeFileSync(TOKEN_FILE, JSON.stringify(tokens, null, 2));
  return oAuth2Client;
}

function getAuthCode(authUrl) {
  return new Promise((resolve, reject) => {
    const server = createServer((req, res) => {
      try {
        const url = new URL(req.url, 'http://127.0.0.1:53682');
        const code = url.searchParams.get('code');
        res.writeHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
        res.end('認証完了です。このタブを閉じてターミナルに戻ってください。');
        server.close();
        if (code) resolve(code);
        else reject(new Error('認可コードがありません'));
      } catch (e) {
        reject(e);
      }
    });
    server.listen(53682, '127.0.0.1', () => {
      console.log('ブラウザで認可してください:\n' + authUrl);
      import('node:child_process').then(({ exec }) => {
        exec(`open "${authUrl}"`);
      });
    });
    // フォールバック: 手動入力
    setTimeout(() => {
      const rl = createInterface({ input: process.stdin, output: process.stdout });
      rl.question('ブラウザが開かない場合、認可コードを貼ってください: ', (code) => {
        rl.close();
        try { server.close(); } catch {}
        resolve(code.trim());
      });
    }, 60000);
  });
}

async function main() {
  const args = parseArgs(process.argv);
  let to = args.to || TO;
  let cc = args.cc !== undefined ? args.cc : CC;
  let subject = args.subject || SUBJECT;
  let body = BODY;

  if (args.test) {
    to = 'pharnewton@gmail.com';
    cc = '';
    subject = `[TEST] ${subject}`;
    body =
      `【テスト送信】本番予定の宛先:\nTo: ${TO}\nCC: ${CC}\nFrom: ${FROM}\n----------\n\n` +
      BODY;
  }

  console.log('=== メール内容 ===');
  console.log(`From: ${FROM}`);
  console.log(`To:   ${to}`);
  console.log(`CC:   ${cc || '(なし)'}`);
  console.log(`Subj: ${subject}`);
  console.log('--- body ---');
  console.log(body);
  console.log('==============');

  if (args['dry-run'] || args.dryrun) {
    console.log('dry-run: API は呼び出していません。');
    return;
  }

  const scopes = args.send ? SCOPES_SEND : SCOPES_DRAFT;
  const auth = await authorize(scopes);
  const gmail = google.gmail({ version: 'v1', auth });
  const raw = buildRawMessage({ to, subject, body, cc: cc || undefined });
  const message = { raw };
  if (args['thread-id']) message.threadId = args['thread-id'];

  if (args.send) {
    const result = await gmail.users.messages.send({ userId: 'me', requestBody: message });
    console.log(`送信完了: id=${result.data.id} threadId=${result.data.threadId}`);
  } else {
    const result = await gmail.users.drafts.create({
      userId: 'me',
      requestBody: { message },
    });
    console.log(
      `下書き作成完了: draftId=${result.data.id} messageId=${result.data.message?.id}`
    );
    console.log('Gmail の下書きを開き、宛先・CC・Zoom URL・ID・パスコードを確認してください。');
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: create_zoom_mail.py
#!/usr/bin/env python3
"""
Zoom URL共有メールを Gmail API で下書き作成（既定）／送信（--send）する。

認証:
  credentials/credentials.json … OAuth クライアント（Desktop 推奨）
  credentials/token.json … 初回認可後に自動保存

スコープ: gmail.compose（下書き） / --send 時は gmail.send も使用
"""

from __future__ import annotations

import argparse
import base64
import os
import sys
from email.message import EmailMessage
from pathlib import Path

from zoom_mail_payload import BODY, CC, FROM, SUBJECT, TO

ROOT = Path(__file__).resolve().parent
CRED_DIR = ROOT / "credentials"
CREDENTIALS_FILE = CRED_DIR / "credentials.json"
TOKEN_FILE = CRED_DIR / "token.json"

SCOPES_DRAFT = ["https://www.googleapis.com/auth/gmail.compose"]
SCOPES_SEND = [
    "https://www.googleapis.com/auth/gmail.compose",
    "https://www.googleapis.com/auth/gmail.send",
]


def build_message(
    *,
    to: str,
    subject: str,
    body: str,
    cc: str | None = None,
    sender: str = FROM,
) -> EmailMessage:
    msg = EmailMessage()
    msg["To"] = to
    msg["From"] = sender
    msg["Subject"] = subject
    if cc:
        msg["Cc"] = cc
    msg.set_content(body)
    return msg


def message_to_raw(msg: EmailMessage) -> dict:
    raw = base64.urlsafe_b64encode(msg.as_bytes()).decode("utf-8")
    return {"raw": raw}


def get_gmail_service(scopes: list[str]):
    try:
        from google.auth.transport.requests import Request
        from google.oauth2.credentials import Credentials
        from google_auth_oauthlib.flow import InstalledAppFlow
        from googleapiclient.discovery import build
    except ImportError as e:
        print(
            "必要なパッケージがありません。次を実行してください:\n"
            "  pip3 install google-api-python-client google-auth-oauthlib",
            file=sys.stderr,
        )
        raise SystemExit(1) from e

    if not CREDENTIALS_FILE.exists():
        print(
            f"OAuth クライアントが見つかりません: {CREDENTIALS_FILE}\n"
            "Google Cloud Console で Desktop アプリの OAuth クライアントを作成し、\n"
            "JSON を credentials/credentials.json として配置してください。\n"
            "必要なスコープ: gmail.compose（および --send 時は gmail.send）",
            file=sys.stderr,
        )
        raise SystemExit(2)

    creds = None
    if TOKEN_FILE.exists():
        creds = Credentials.from_authorized_user_file(str(TOKEN_FILE), scopes)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                str(CREDENTIALS_FILE), scopes
            )
            creds = flow.run_local_server(port=0)
        CRED_DIR.mkdir(parents=True, exist_ok=True)
        TOKEN_FILE.write_text(creds.to_json(), encoding="utf-8")

    return build("gmail", "v1", credentials=creds)


def create_draft(service, msg: EmailMessage, thread_id: str | None = None) -> dict:
    body = {"message": message_to_raw(msg)}
    if thread_id:
        body["message"]["threadId"] = thread_id
    return service.users().drafts().create(userId="me", body=body).execute()


def send_message(service, msg: EmailMessage, thread_id: str | None = None) -> dict:
    raw = message_to_raw(msg)
    if thread_id:
        raw["threadId"] = thread_id
    return service.users().messages().send(userId="me", body=raw).execute()


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Zoom共有メールの下書き作成／送信")
    p.add_argument(
        "--send",
        action="store_true",
        help="下書きではなく送信する（確認後のみ使用）",
    )
    p.add_argument(
        "--test",
        action="store_true",
        help="テスト宛先 pharnewton@gmail.com へ（本番 To/CC は使わない）",
    )
    p.add_argument("--to", default=None, help="宛先上書き")
    p.add_argument("--cc", default=None, help="CC上書き")
    p.add_argument("--subject", default=None, help="件名上書き")
    p.add_argument("--thread-id", default=None, help="既存スレッドへ紐付け")
    p.add_argument(
        "--dry-run",
        action="store_true",
        help="APIを呼ばず内容だけ表示する",
    )
    return p.parse_args()


def main() -> int:
    args = parse_args()

    to = args.to or TO
    cc = args.cc if args.cc is not None else CC
    subject = args.subject or SUBJECT
    body = BODY

    if args.test:
        to = "pharnewton@gmail.com"
        cc = ""
        subject = f"[TEST] {subject}"
        body = (
            "【テスト送信】本番予定の宛先:\n"
            f"To: {TO}\n"
            f"CC: {CC}\n"
            f"From: {FROM}\n"
            "----------\n\n"
            + BODY
        )

    msg = build_message(to=to, subject=subject, body=body, cc=cc or None)

    print("=== メール内容 ===")
    print(f"From: {FROM}")
    print(f"To:   {to}")
    print(f"CC:   {cc or '(なし)'}")
    print(f"Subj: {subject}")
    print("--- body ---")
    print(body)
    print("==============")

    if args.dry_run:
        print("dry-run: API は呼び出していません。")
        return 0

    scopes = SCOPES_SEND if args.send else SCOPES_DRAFT
    service = get_gmail_service(scopes)

    if args.send:
        result = send_message(service, msg, args.thread_id)
        print(f"送信完了: id={result.get('id')} threadId={result.get('threadId')}")
    else:
        result = create_draft(service, msg, args.thread_id)
        draft_id = result.get("id")
        message = result.get("message") or {}
        print(
            f"下書き作成完了: draftId={draft_id} "
            f"messageId={message.get('id')} threadId={message.get('threadId')}"
        )
        print("Gmail の下書きを開き、宛先・CC・Zoom URL・ID・パスコードを確認してください。")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
